### PR TITLE
Adapt for sphinx 2.x

### LINKED ISSUE #55: Support Sphinx 2+
`sphinxcontrib-images` has `sphinx>=1.1.3,<2` as dependency - but as sphinx 2.0.0 has recently been released (03/28), this broke our builds - version conflict, as we install sphinx with open ended version currently.

### PR DESCRIPTION
**Pull request by t-b**
This is an initial port to sphinx 2 and therefore also to python 3.

It is not yet working as I get the following errors on running the tests:

```
Exception occurred:
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] Datei oder Verzeichnis nicht gefunden: '/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib_images_lightbox2/lightbox2/css/lightbox.css'
The full traceback has been saved in /tmp/sphinx-err-ou2pkbz7.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

```
$cat /tmp/sphinx-err-ou2pkbz7.log
# Sphinx version: 2.1.0
# Python version: 3.6.7 (CPython)
# Docutils version: 0.14 
# Jinja2 version: 2.10.1
# Last messages:
#   https://upload.wikimedia.org/wikipedia/meta/0/08/Wikipedia-logo-v2_1x.png -> /home/firma/devel/sphinxcontrib-images/docs/source/_images/f453ce6fa37bfa485a753fc8b2286dee1df0f620 (already in cache)
#   
#   Downloading remote images...[100%] http://a/non_existing_image.png
#   
#   http://a/non_existing_image.png -> /home/firma/devel/sphinxcontrib-images/docs/source/_images/fb93720819f0fd7bb5da5e1596197fb67d90bf1b (already in cache)
#   
#   
#   
#   Copying static files for sphinxcontrib-images...[ 11%] lightbox2/css/lightbox.css
#   
# Loaded extensions:
#   sphinx.ext.mathjax (2.1.0) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.1) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.1) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (1.0.2) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.3) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.2) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc (2.1.0) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py
#   sphinx.ext.todo (2.1.0) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/ext/todo.py
#   sphinx.ext.viewcode (2.1.0) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/ext/viewcode.py
#   sphinxcontrib.images (2.1.0) from /home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/images.py
Traceback (most recent call last):
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/cmd/build.py", line 284, in build_main
    app.build(args.force_all, filenames)
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/application.py", line 345, in build
    self.builder.build_update()
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 319, in build_update
    len(to_build))
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 332, in build
    updated_docnames = set(self.read())
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 444, in read
    for retval in self.events.emit('env-updated', self.env):
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib/images.py", line 241, in install_backend_static_files
    copyfile(source_file_path, dest_file_path)
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinx/util/osutil.py", line 130, in copyfile
    shutil.copyfile(source, dest)
  File "/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] Datei oder Verzeichnis nicht gefunden: '/home/firma/devel/sphinxcontrib-images/.tox/py36/lib/python3.6/site-packages/sphinxcontrib_images_lightbox2/lightbox2/css/lightbox.css'
```

Any help is appreciated.

Close #55.

### CODE REVIEW

**Comment by jonascj** [comment]
Great start I hope!
So far I've figured out this error has to do with https://github.com/spinus/sphinxcontrib-images/tree/master/sphinxcontrib_images_lightbox2 which references a git submodule (the lightbox2 files). They appear not to be cloned during the tests, resulting in ``sphinxcontrib-images/sphinxcontrib_images_lightbox2/lightbox`` being empty, hence it fails to copy anything from that directory to, also in this warning: ``warning: no files found matching 'sphinxcontrib_images_lightbox2/lightbox2/css/lightbox.css`` and in the end it fails to find anything at ``FileNotFoundError: [Errno 2] No such file or directory: '/tmp/sphinxcontrib-images/.tox/py37-sphinx21/lib/python3.7/site-packages/sphinxcontrib_images_lightbox2/lightbox2/css/lightbox.css``.

**Comment by jonascj** [comment]
>   py37-sphinx20: commands succeeded
>   py37-sphinx21: commands succeeded
If you manually change directory to ``sphinxcontrib-images/sphinxcontrib_images_lightbox2`` and clone the lightbox2 repo ``git clone https://github.com/lokesh/lightbox2.git`` the tests succeeds for the two interpreters I have installed (or something along those lines).

**Comment by jonascj** [comment]
So in the end it just turns out both @t-b and I forgot to clone the ``sphinxcontrib-imaegs`` repo including submodules :-$

So to successfully pass the tests clone as:

```
git clone --recurse-submodules https://github.com/spinus/sphinxcontrib-images.gi
```

And pull this PR:

```
git pull origin pull/56/hea
```

So what now, since the tests pass? Test it with a real project, I'll do that later today if I can figure out how.

**Comment by jonascj** [comment]
Just tested it on one of my projects, previously built with Sphinx 1.8.x and sphinxcontrib-images.
The project seemingly builds just fine with the version of sphinxcontrib-images in this PR and Sphinx 2.1.2.


```
git clone --recurse-submodules https://github.com/spinus/sphinxcontrib-images.git
git pull origin pull/56/head
cd sphinxcontrib-images
# make a virtual environment however you like
pip install sphinx
pip install -e .
# change directory to your sphinx project
make html
```
@spinus Should we/you/someone move forward with merging this and bumping the release on pypi?

**Comment by spinus** [comment]
@jonascj thank you for pushing this forward. I think we should but I'm not sure I would like to drop the support completely for older versions like this PR does.
Let me review the code and if you have time, maybe we could work through the week and merge/release at the end?

File: setup.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def run_tests(self):
     packages=find_packages(),
     include_package_data=True,
     setup_requires=['wheel'],
-    install_requires=['sphinx>=1.1.3,<2',
```
Comment (spinus) [inline setup.py]: if possible I wouldn't drop support to exsiting versions, unless that's completely necessary.

File: sphinxcontrib/images.py [hunk 2]
```diff
@@ -92,8 +95,8 @@ def directive_boolean(value):
     elif value.lower().strip() in ['no', '0', 0, 'false', 'none']:
         return False
     else:
-        raise ValueError(u"Please use on of: yes, true, no, false. "
-                         u"Do not use `{}` as boolean.".format(value))
+        raise ValueError("Please use on of: yes, true, no, false. "
```
Comment (spinus) [inline sphinxcontrib/images.py]: I'm not sure why it's here, that's nice support for both py2 and py3, no?

File: sphinxcontrib/images.py [hunk 3]
```diff
@@ -151,7 +154,7 @@ def run(self):
         #TODO: something is broken here, not parsed as expected
         description = nodes.paragraph()
         content = nodes.paragraph()
-        content += [nodes.Text(u"%s" % x) for x in self.content]
+        content += [nodes.Text("%s" % x) for x in self.content]
```
Comment (spinus) [inline sphinxcontrib/images.py]: same here ^^

File: sphinxcontrib/images.py [hunk 4]
```diff
@@ -252,20 +255,20 @@ def download_images(app, env):
             'Downloading remote images...',
             brown,
             len(env.remote_images)):
-                
+
         dst = os.path.join(env.srcdir, env.remote_images[src])
         if not os.path.isfile(dst):
-            app.info('{} -> {} (downloading)'
+            logger.info('{} -> {} (downloading)'
```
Comment (spinus) [inline sphinxcontrib/images.py]: that's how logging was done in sphinx, I'm not up to date, but in sphinx2 should that be changed to logging? is moving to logging compatible in sphinx1?

File: sphinxcontrib/images.py [hunk 5]
```diff
@@ -365,6 +368,6 @@ def main(args=sys.argv[1:]):
         backends = pkg_resources.iter_entry_points(group='sphinxcontrib.images.backend')
         if backends:
             for backend in backends:
-                print ('- {0.name} (from `{0.dist}` package)'.format(backend))
+                print(('- {0.name} (from `{0.dist}` package)'.format(backend)))
```
Comment (spinus) [inline sphinxcontrib/images.py]: +1

**Comment by spinus** [comment]
I'll try to add submodules update on travis later

File: sphinxcontrib/images.py (hunk 4 above)
Comment (jonascj) [reply sphinxcontrib/images.py]:
```
from sphinx.util import logging
logger = logging.getLogger(__name__)
```

can be found in many files of the sphinx project, for both 1.6, 1.7, 1.8 and 2.x tags, e.g.:

Not here: https://github.com/sphinx-doc/sphinx/blob/v1.5.6/sphinx/application.py
But here: https://github.com/sphinx-doc/sphinx/blob/v1.6/sphinx/application.py
And here: https://github.com/sphinx-doc/sphinx/blob/v2.1.2/sphinx/application.py

So it seems compatible for 1.6 or newer if I am any judge.

File: sphinxcontrib/images.py (hunk 2 above)
Comment (jonascj) [reply sphinxcontrib/images.py]: I'm not sure I follow your comment. Why is what where? 

``s=u"Sömething ûnicode"`` is both valid in python3 and python2 afaik. In python2  s would have type ``unicode`` in python3 just ``str``.

File: sphinxcontrib/images.py (hunk 3 above)
Comment (jonascj) [reply sphinxcontrib/images.py]: So we could keep the ``u"string"`` style for compatibility, is that your thought?

File: sphinxcontrib/images.py (hunk 5 above)
Comment (jonascj) [reply sphinxcontrib/images.py]: Unnecessary.

File: setup.py (hunk 1 above)
Comment (jonascj) [reply setup.py]: I have no idea what is possible and what is not. The old requirement of >=1.1.3 and <2 already spans versions where sphinx seemingly switched from method of logging to another (between 1.5 and 1.6 there appears to be a change in logging).

According to this https://github.com/sphinx-doc/sphinx/issues/4375#issuecomment-355316362 (and likely other places as well) Sphinx 2.0 and forward will be Python3 only. So likely 1.7 and 1.8 were for both python2 and python3. 

Maybe it is only worth supporting back to releases supporting python3? People can still install an old version if they really need this extension with sphinx 1.1 ...

**Comment by jonascj** [comment]
> @jonascj thank you for pushing this forward. I think we should but I'm not sure I would like to drop the support completely for older versions like this PR does.
> Let me review the code and if you have time, maybe we could work through the week and merge/release at the end?

I'd be cool to have a new release within 5-10 days. Let me know if I can do anything to help. I have no experience with travis, I just figured out the missed cloning of the submodule locally.

It has also been a long time since i worked on anything python2, so I am not sure when sphinx was a pure python2 project. Supporting versions that old seems hard. Sphinx 2.0+ seems to be python3 only, so if this extension becomes python3 only as well then you could possibly support back until sphinx did not support python3...

**Comment by jonascj** [comment]
@spinus I don't know if this means anything, but I tried to run the tests with the following changes versions in ``tox.ini`` :
```
[tox]
envlist = {py34,py35,py36,py37,py38,pypy35}-sphinx{11,17,18,20,21}

{ removed content }

deps =
    pip
    setuptools
    wheel
    sphinx-rtd-theme==0.1.6
    sphinx20: sphinx==2.0
    sphinx21: sphinx==2.1
    sphinx18: sphinx==1.8
    sphinx17: sphinx==1.7
    sphinx10: sphinx==1.1.3
```

and these requirements in setup.py:

```
install_requires=['sphinx>=1.1.3',                                          
                           'requests>2.2,<3']
```

With python 3.7 these tests pass for this PR:
```
  py37-sphinx11: commands succeeded
  py37-sphinx17: commands succeeded
  py37-sphinx18: commands succeeded
  py37-sphinx20: commands succeeded
  py37-sphinx21: commands succeeded
```

But I am not even remotely sure that means this PR-version will work with those versions of sphinx.

**Comment by t-b** [comment]
@jonascj Thanks for the submodule hint. This was really the source of the problem.

I've removed the `2to3` script changes. And also kept the code compatible with older sphinx versions.

I've started adapting the travis config but some stuff is still failing, also the environment marker in setup.py are somehow not working as documented.

**Comment by t-b** [comment]
@jonascj @spinus Done and therefore ready for review.

**Comment by t-b** [comment]
@spinus Friendly ping.

**Comment by t-b** [comment]
pong.

**Comment by jonascj** [comment]
@t-b I am not sure how to initiate the review etc. Is it @spinus who has to initiate it?

**Comment by t-b** [comment]
@jonascj We can only wait.

**Comment by spinus** [comment]
hello there, sorry guys, I have a hard time now. Anybody I could transfer this project to and PyPI access to the package? I know there is a way as well to transfer that to sphinxcontrib stuff but also not time for that. Please let me know.

**Comment by jonascj** [comment]
Sorry to hear it.
https://github.com/sphinx-contrib is a collection of extensions, some of them orphaned by the original author but taken over by others. I don't know how to get there, maybe @dhellmann or @shimizukawa from that repo knows.

**Comment by dhellmann** [comment]
IIRC, someone with admin rights on this repository needs to join the sphinx-contrib org on GitHub. Then it will be possible for that person to transfer the repository, with an org admin approving the request. I will approve the transfer when you're ready.

It sounds like @spinus doesn't have much time right now, so maybe someone else with admin rights can initiate the transfer?

**Comment by spinus** [comment]
@dhellmann thank you for help. I did the transfer but it told me "You don’t have the permission to create public repositories on sphinx-contrib". Other way would be to do a fork and I'll put the message and links to new repo.

**Comment by spinus** [comment]
@jonascj thank you for stepping in, appreciate a lot.

**Comment by dhellmann** [comment]
> @dhellmann thank you for help. I did the transfer but it told me "You don’t have the permission to create public repositories on sphinx-contrib". Other way would be to do a fork and I'll put the message and links to new repo.

Oops. Maybe I have the permissions incomplete (it has been a while since I've done this).

If you can give me admin rights here, I should be able to move it for you.

**Comment by dhellmann** [comment]
I have push access here now, but that's not sufficient to let me see the button in the Settings to move the repo to a new org.

**Comment by spinus** [comment]
@dhellmann right. I couldn't grant admin at first, I had to wait to accept invitation. Changing now.

**Comment by spinus** [comment]
image
hmm, looks like I don't have usual option to change permission type :/ maybe forking is better way?

**Comment by dhellmann** [comment]
I forked the repo into the sphinx-contrib org and added @spinus with admin access there.

Unfortunately that means @t-b (or someone) will need to re-target this PR to the new repo.

If anyone else would like write access to help manage the code, please open an issue on the new repo and either @spinus or I can take care of it.

**Comment by t-b** [comment]
I'm also fine with a fork at https://github.com/sphinx-contrib/images. Just remember to archive this repo so that users know that this is not active anymore.

**Comment by jonascj** [comment]
@t-b so will you make a new PR to the new repo?

@spinus should we request some more admins for the project, e.g. t-b and myself?

**Comment by t-b** [comment]
@jonascj I've opened https://github.com/sphinx-contrib/images/pull/1.

**Comment by t-b** [comment]
@jonascj  I might be able to occasionally chim in to help.

**Comment by spinus** [comment]
@jonascj sure, if you have time and will to push things forward, go for it!

**Comment by spinus** [comment]
@dhellmann should I migrate PyPI package as well somehow? Maybe I can write to mods there and they can switch "org"?

**Comment by jonascj** [comment]
@spinus Not that I know the first thing about releasing packages to PyPI or building releases of this package, but the goal must be to have the Images for Sphinx 2.x be available via pip (PyPI). If we continue development at https://github.com/sphinx-contrib/ I suppose https://pypi.org/project/sphinxcontrib-images/ should point to that repo instead of this repo.

Maybe this repo should have it's readme updated to tell of the new location ...

I'm not sure what the package naming convention is for the projects/packages hosted in https://github.com/sphinx-contrib/ . Should it still be `pip install sphinxcontrib-images` or something else?

**Comment by spinus** [comment]
@jonascj sure thing, but both readme and whl/zip files were uploaded by my, by hand. So changing repo there won't change much, I still need to pass credentials/ownership of this, so somebody can do (or setup automation) to release there.

**Pull request closed without merging**